src/lib.rs

```diff
@@ -11,6 +11,678 @@
  * SPDX-License-Identifier: Apache-2.0
  ********************************************************************************/
 
+use std::{
+    collections::{HashMap, HashSet},
+    str::FromStr,
+    sync::Arc,
+};
+
+use async_std::{sync::RwLock, task::block_on};
+use log::info;
+use mqtt::AsyncClient;
+use paho_mqtt::{self as mqtt, MQTT_VERSION_5};
+use protobuf::Enum;
+use up_rust::{ComparableListener, UAttributes, UCode, UMessage, UPriority, UStatus, UUri, UUID};
+
+pub mod rpc;
 pub mod transport;
 
-pub struct UPClientMqtt {}
+/// Constants defining the protobuf field numbers for UAttributes.
+// TODO: Convert this to a dynamically generated list.
+pub const ID_NUM: &str = "1";
+pub const TYPE_NUM: &str = "2";
+pub const SOURCE_NUM: &str = "3";
+pub const SINK_NUM: &str = "4";
+pub const PRIORITY_NUM: &str = "5";
+pub const TTL_NUM: &str = "6";
+pub const PERM_LEVEL_NUM: &str = "7";
+pub const COMMSTATUS_NUM: &str = "8";
+pub const REQID_NUM: &str = "9";
+pub const TOKEN_NUM: &str = "10";
+pub const TRACEPARENT_NUM: &str = "11";
+pub const PAYLOAD_NUM: &str = "12";
+
+// URI Wildcard consts
+const WILDCARD_AUTHORITY: &str = "*";
+const WILDCARD_ENTITY_ID: u32 = 0x0000_FFFF;
+const WILDCARD_ENTITY_VERSION: u32 = 0x0000_00FF;
+const WILDCARD_RESOURCE_ID: u32 = 0x0000_FFFF;
+
+/// Configuration for the mqtt client.
+pub struct MqttConfig {
+    /// Port of the mqtt broker to connect to.
+    pub mqtt_port: String,
+    /// Hostname of the mqtt broker.
+    pub mqtt_hostname: String,
+    /// Optional SSL options for the mqtt connection.
+    pub ssl_options: Option<mqtt::SslOptions>,
+}
+
+/// UP Client for mqtt.
+pub struct UPClientMqtt {
+    mqtt_client: Arc<mqtt::AsyncClient>,
+    topic_listener_map: Arc<RwLock<HashMap<String, HashSet<ComparableListener>>>>,
+    // My authority
+    authority_name: String,
+    // Indicates where client instance is running.
+    client_type: UPClientMqttType,
+}
+
+pub enum UPClientMqttType {
+    Device,
+    Cloud,
+}
+
+#[allow(dead_code)]
+impl UPClientMqtt {
+    /// Create a new UPClientMqtt.
+    ///
+    /// # Arguments
+    /// * `config` - Configuration for the mqtt client.
+    /// * `client_id` - Client id for the mqtt client.
+    pub async fn new(
+        config: MqttConfig,
+        client_id: UUID,
+        authority_name: String,
+        client_type: UPClientMqttType,
+    ) -> Result<Self, UStatus> {
+        let topic_listener_map = Arc::new(RwLock::new(HashMap::new()));
+
+        let ptr = topic_listener_map.clone();
```

Comment:
Naming of ptr? Seems a little unintuitive. I don't love it, but often go with varname_clone.

Reply:
how about 'topic_listener_map_handle' - not amazing either

Reply:
Yeah, but still probably a bit more descriptive 🙂

Fine by me